File: passman-core/src/crypto.rs
use aes_gcm::{
    aead::{Aead, AeadCore, KeyInit, OsRng},
    Aes256Gcm, Nonce,
};
use argon2::{Argon2, Params};
use rand::RngCore;
use thiserror::Error;

pub const KEY_SIZE: usize = 32;
pub const NONCE_SIZE: usize = 12;
pub const TAG_SIZE: usize = 16;
pub const SALT_SIZE: usize = 16;

#[derive(Debug, Error)]
pub enum CryptoError {
    #[error("argon2 error: {0}")]
    Argon2(String),
    #[error("encryption error: {0}")]
    Encrypt(String),
    #[error("incorrect password")]
    Decrypt,
}

#[derive(Debug, Clone)]
pub struct KdfParams {
    pub salt: [u8; SALT_SIZE],
    pub memory_kb: u32,
    pub iterations: u32,
    pub parallelism: u32,
}

impl Default for KdfParams {
    fn default() -> Self {
        SecurityLevel::Medium.kdf_params()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SecurityLevel {
    Low,
    Medium,
    Secure,
    Best,
}

impl SecurityLevel {
    pub fn kdf_params(self) -> KdfParams {
        let salt: [u8; SALT_SIZE] = random_bytes(SALT_SIZE)
            .try_into()
            .expect("salt generated with correct length");
        let (memory_kb, iterations, parallelism) = match self {
            SecurityLevel::Low => (32_768, 2, 2),
            SecurityLevel::Medium => (65_536, 3, 4),
            SecurityLevel::Secure => (131_072, 4, 4),
            SecurityLevel::Best => (262_144, 6, 8),
        };
        KdfParams {
            salt,
            memory_kb,
            iterations,
            parallelism,
        }
    }

    pub fn label(self) -> &'static str {
        match self {
            SecurityLevel::Low => "low",
            SecurityLevel::Medium => "medium",
            SecurityLevel::Secure => "secure",
            SecurityLevel::Best => "best",
        }
    }
}

impl std::str::FromStr for SecurityLevel {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "low" => Ok(Self::Low),
            "medium" => Ok(Self::Medium),
            "secure" => Ok(Self::Secure),
            "best" => Ok(Self::Best),
            other => Err(format!("unknown security level: {other}")),
        }
    }
}

pub fn random_bytes(len: usize) -> Vec<u8> {
    let mut buf = vec![0u8; len];
    OsRng.fill_bytes(&mut buf);
    buf
}

pub fn derive_key(password: &str, params: &KdfParams) -> Result<[u8; KEY_SIZE], CryptoError> {
    let argon2 = Argon2::new(
        argon2::Algorithm::Argon2id,
        argon2::Version::V0x13,
        Params::new(
            params.memory_kb,
            params.iterations,
            params.parallelism,
            Some(KEY_SIZE),
        )
        .map_err(|e| CryptoError::Argon2(e.to_string()))?,
    );

    let mut key = [0u8; KEY_SIZE];
    argon2
        .hash_password_into(password.as_bytes(), &params.salt, &mut key)
        .map_err(|e| CryptoError::Argon2(e.to_string()))?;
    Ok(key)
}

/// Result of an AES-256-GCM encryption: the random nonce and the ciphertext bytes.
pub struct Ciphertext {
    pub nonce: Vec<u8>,
    pub bytes: Vec<u8>,
}

pub fn encrypt(plaintext: &[u8], key: &[u8; KEY_SIZE]) -> Ciphertext {
    let cipher = Aes256Gcm::new(key.into());
    let nonce = Aes256Gcm::generate_nonce(&mut OsRng);
    let bytes = cipher
        .encrypt(&nonce, plaintext)
        .expect("encryption should not fail");
    Ciphertext {
        nonce: nonce.to_vec(),
        bytes,
    }
}

pub fn decrypt(
    ciphertext: &[u8],
    key: &[u8; KEY_SIZE],
    nonce: &[u8],
) -> Result<Vec<u8>, CryptoError> {
    let cipher = Aes256Gcm::new(key.into());
    let nonce = Nonce::from_slice(nonce);
    cipher
        .decrypt(nonce, ciphertext)
        .map_err(|_| CryptoError::Decrypt)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_derive_key() {
        let params = KdfParams::default();
        let key1 = derive_key("password", &params).unwrap();
        let key2 = derive_key("password", &params).unwrap();
        let key3 = derive_key("different", &params).unwrap();
        assert_eq!(key1, key2);
        assert_ne!(key1, key3);
    }

    #[test]
    fn test_encrypt_decrypt() {
        let key = random_bytes(KEY_SIZE);
        let key: &[u8; KEY_SIZE] = key.as_slice().try_into().unwrap();
        let plaintext = b"hello world";
        let encrypted = encrypt(plaintext, key);
        let decrypted = decrypt(&encrypted.bytes, key, &encrypted.nonce).unwrap();
        assert_eq!(plaintext.to_vec(), decrypted);
    }
}

File: pkg/crypto/crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"

	"golang.org/x/crypto/argon2"
)

const (
	KeySize   = 32
	NonceSize = 12
	TagSize   = 16
	SaltSize  = 16
)

type SecurityLevel string

const (
	SecurityLevelLow    SecurityLevel = "low"
	SecurityLevelMedium SecurityLevel = "medium"
	SecurityLevelSecure SecurityLevel = "secure"
	SecurityLevelBest   SecurityLevel = "best"
)

func ParseSecurityLevel(s string) (SecurityLevel, error) {
	switch s {
	case "low":
		return SecurityLevelLow, nil
	case "medium":
		return SecurityLevelMedium, nil
	case "secure":
		return SecurityLevelSecure, nil
	case "best":
		return SecurityLevelBest, nil
	default:
		return "", fmt.Errorf("unknown security level: %s", s)
	}
}

func (l SecurityLevel) Label() string {
	return string(l)
}

func (l SecurityLevel) KdfParams() KdfParams {
	salt := RandomBytes(SaltSize)
	var memoryKB, iterations, parallelism uint32
	switch l {
	case SecurityLevelLow:
		memoryKB, iterations, parallelism = 32768, 2, 2
	case SecurityLevelMedium:
		memoryKB, iterations, parallelism = 65536, 3, 4
	case SecurityLevelSecure:
		memoryKB, iterations, parallelism = 131072, 4, 4
	case SecurityLevelBest:
		memoryKB, iterations, parallelism = 262144, 6, 8
	}
	return KdfParams{
		Salt:       [SaltSize]byte(salt),
		MemoryKB:   memoryKB,
		Iterations: iterations,
		Parallelism: parallelism,
	}
}

type KdfParams struct {
	Salt        [SaltSize]byte
	MemoryKB    uint32
	Iterations  uint32
	Parallelism uint32
}

func DefaultKdfParams() KdfParams {
	return SecurityLevelMedium.KdfParams()
}

func RandomBytes(n int) []byte {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		panic(err)
	}
	return b
}

func DeriveKey(password string, params *KdfParams) ([KeySize]byte, error) {
	key := argon2.IDKey(
		[]byte(password),
		params.Salt[:],
		params.Iterations,
		params.MemoryKB,
		uint8(params.Parallelism),
		KeySize,
	)
	var result [KeySize]byte
	copy(result[:], key)
	return result, nil
}

type Ciphertext struct {
	Nonce []byte
	Bytes []byte
}

func Encrypt(plaintext []byte, key *[KeySize]byte) Ciphertext {
	block, err := aes.NewCipher(key[:])
	if err != nil {
		panic(err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		panic(err)
	}
	nonce := make([]byte, gcm.NonceSize())
	if _, err := rand.Read(nonce); err != nil {
		panic(err)
	}
	encrypted := gcm.Seal(nil, nonce, plaintext, nil)
	return Ciphertext{Nonce: nonce, Bytes: encrypted}
}

func Decrypt(ciphertext []byte, key *[KeySize]byte, nonce []byte) ([]byte, error) {
	block, err := aes.NewCipher(key[:])
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, errors.New("incorrect password")
	}
	return plaintext, nil
}

// KdfParamsJSON is the JSON-serializable form of KdfParams (salt is base64-encoded).
type KdfParamsJSON struct {
	Salt        string `json:"salt"`
	MemoryKB    uint32 `json:"memory_kb"`
	Iterations  uint32 `json:"iterations"`
	Parallelism uint32 `json:"parallelism"`
}

func KdfParamsToJSON(p *KdfParams) KdfParamsJSON {
	return KdfParamsJSON{
		Salt:        base64.StdEncoding.EncodeToString(p.Salt[:]),
		MemoryKB:    p.MemoryKB,
		Iterations:  p.Iterations,
		Parallelism: p.Parallelism,
	}
}

func KdfParamsFromJSON(j *KdfParamsJSON) (KdfParams, error) {
	saltBytes, err := base64.StdEncoding.DecodeString(j.Salt)
	if err != nil {
		return KdfParams{}, err
	}
	if len(saltBytes) != SaltSize {
		return KdfParams{}, errors.New("invalid salt length")
	}
	var p KdfParams
	copy(p.Salt[:], saltBytes)
	p.MemoryKB = j.MemoryKB
	p.Iterations = j.Iterations
	p.Parallelism = j.Parallelism
	return p, nil
}

// MarshalJSON for KdfParams so it can be embedded directly if needed.
func (p KdfParams) MarshalJSON() ([]byte, error) {
	return json.Marshal(KdfParamsToJSON(&p))
}

// UnmarshalJSON for KdfParams.
func (p *KdfParams) UnmarshalJSON(data []byte) error {
	var j KdfParamsJSON
	if err := json.Unmarshal(data, &j); err != nil {
		return err
	}
	params, err := KdfParamsFromJSON(&j)
	if err != nil {
		return err
	}
	*p = params
	return nil
}
